app/Kconfig

```diff
@@ -301,6 +301,24 @@ config ZMK_BACKLIGHT_AUTO_OFF_USB
 #ZMK_BACKLIGHT
 endif
 
+config ZMK_ANIMATION
+    bool "RGB Animations"
+    select LED_STRIP
+
+if ZMK_ANIMATION
+
+config ZMK_ANIMATION_FPS
+    int "Animation FPS"
+    range 0 60
```

Comment:
Wouldn't we want `range 1 60` here, since 0 FPS doesn't make sense?

Reply:
Does having a global animation speed even make sense? Maybe we could have a global *default* in Kconfig like we do for other features, but I would expect for different animations to potentially want to run at different speeds.

Reply:
(Also, 0 FPS does make sense for something like a static color, but then I question the need for an FPS property on that animation to begin with)

Reply:
(Disclaimer: I haven't yet had time to look at the code for this yet)

If you look at earlier conversations on this PR, I think we identified two different ways that animations may need to update:

1. At a steady framerate (e.g. rainbow wave)
2. Event-based (e.g. caps lock indicator)

One way I've seen this type of thing implemented which seemed to work well in another project was to have a function which can be called to request a render (handles event-based animations) and then the render callback returns a value indicating whether it wants to request another frame. If it returns true, then the animation continues. If it returns false, then the animation stops and waits for a call to the request render function.

If we implement things like that, then 0 FPS indeed does not make sense, because static animations would just be event-based.

Reply:
Yeah, so it mostly works like @joelspadin mentioned. The animations are responsible for calling `zmk_animation_request_frames` when their state is updated. Unless an animation requested frames, the entire system becomes static until the next request. It's best to think of it as a limit rather than how many frames are actually processed.

But it is possible for animations to request more than one frame at a time, and that's where this config variable comes in. That's the way the ripple animation example is implemented, where it will immediately request a number of frames necessary to complete the entire effect animation.

Calls to `zmk_animation_request_frames` need to be throttled in case there's multiple animations reacting to the same event (for example the said caps lock indicator combined with the ripple effect).  So even if we remove the config var itself, we'd just need to hardcode a number anyway in order to be able to aggregate updates like this at the top level.

As for running different animations at different speeds, if you add such a setting on the animation, it can totally be done with the current system without any compromises. But obviously it would increase the complexity of the animation itself by quite a bit so... The door is open, if anybody deems it worth the effort.

Finally, I do agree `0` likely doesn't make sense and I'm not sure it'd even compile 🤔. I'm looking at it now and all I see is a bunch of potential division by `0` issues. I'll change that to `1`.

Reply:
I'm still unsure how much sense it makes to allow animations to request multiple frames at once. I can think of three cases:

1. The animation doesn't need to update more than once per event, so requesting multiple frames isn't needed.
2. The animation runs forever, so it needs to request N new frames every N frames. N can be 1, so requesting multiple frames isn't needed.
3. The animation starts on an event and then runs for a set number of frames. This is where requesting multiple frames is potentially useful, however I would expect that the render function has to track its progress through the animation anyways, so it needing to return a value indicating whether it's done or not shouldn't make it any more complex.

Are there other cases I'm missing?

Reply:
Yeah, that's another way to do it. I guess I never thought about it as a return value though... I was only thinking about calling the equivalent of `zmk_animation_request_frames(1)` until done, like you also mentioned.  
I wasn't a fan of that approach (although obviously you could still do it with this API), as there could be some potential for race conditions.

But you're right, if we just return a value it should be all right. I'll see if I can draft that up.

Reply:
One thing to consider if we implement rendering this way is how we handle "events". Currently, each animation driver can implement "events" however it pleases- IE a ripple animation can utilize a key event, or a periodic animation could request frames on a timer.

If animations are instead notified about events, I am concerned about the flexibility of events. We would likely need to pass every key event to every animation, unless we have a way for animations to notify the root animation driver what events are relevant to them.

> but I would expect for different animations to potentially want to run at different speeds.

I'm not sure this makes sense. Currently, the way the animation subsystem is written is that each animation driver receives what is effectively a pixel buffer, and is permitted to update pixels however it sees fit (including implementing blending effects). If animations run at a different framerate, we loose this ability. I think that static animations might be a case we need to handle differently, but if an animation is dynamic I think it makes sense for it to be rendered at a global framerate.

Reply:
I wasn't suggesting that different animations would be allowed to run asynchronously at different framerates. If everything goes through a single, root animation, you could have a driver API function that returns the framerate for that animation. An animation that composes other animations could then return the maximum of the child animations' framerates. This does mean that animations would need to track their progress using time elapsed rather than frame counts though, because they could be running faster than requested. We could maybe start with a global framerate and then add a way to adjust the framerate dynamically later if it would be useful. 

I would expect animation drivers to register for events on their own. There is one potential problem with this in that ZMK doesn't allow for dynamically registering and unregistering event handlers currently, so we would need to add a way to track which devices want to receive events.

More Python-ish pseudocode for the design I had in mind:

```py
class LoopingAnimation:
    def activate(self):
        zmk_animation_request_frame()

    def deactivate(self):
        pass

    def get_framerate(self):
        return self.framerate

    def render(self, ...):
        ...
        return True

class RippleAnimation:
    def activate(self):
        zmk_event_register(zmk_position_state_changed, self.key_changed)

    def deactivate(self):
        zmk_event_unregister(zmk_position_state_changed, self.key_changed)

    def key_changed(self, ev):
        self.ripple_duration[ev.position] = 100
        zmk_animation_request_frame()

    def get_framerate(self, ...):
        return self.framerate

    def render(self, ...):
        ...
        return any(duration > 0 for duration in self.ripple_duration)

class CompositeAnimation:
    def activate(self):
        for anim in self.animations:
            anim.activate()

    def deactivate(self):
        for anim in self.animations:
            anim.deactivate()

    def get_framerate(self, ...):
        return max(anim.get_framerate() for anim in self.animations)

    def render(self, ...):
        animating = False
        for anim in self.animations:
            child_animating = anim.render(self, ...)
            animating = animating or child_animating
        return animating

class AnimationSelector:
    def activate(self):
        self.animations[self.index].activate()

    def deactivate(self):
        self.animations[self.index].deactivate()

    def get_framerate(self):
        return self.animations[self.index].get_framerate()

    def render(self):
        return self.animations[self.index].render(...)

    def set_animation(self, index):
        if index != self.index:
            self.animations[self.index].deactivate()
            self.index = index
            self.animations[self.index].activate()
```

Since ZMK doesn't support dynamically registering/unregistering events, we would need to do something like

```c
// Maintain a list of devices that are registered for this event
static struct device* listeners[DRIVER_COUNT];
static int listeners_len = 0;
static void add_listener(struct device* device) { ... }
static void remove_listener(struct device* device) { ... }

// Handle the event globally, and forward to each device that is registered
static int ripple_animation_listener(const zmk_event_t* ev) {
    const struct zmk_position_state_changed* data;
    if ((data = as_zmk_position_state_changed(ev)) != NULL) {
        for (int i = 0; i < listeners_len; i++) {
            ripple_animation_position_state_changed(listeners[i], data);
        }
    }
}

ZMK_LISTENER(ripple_animation, ripple_animation_listener);
ZMK_EVENT_SUBSCRIPTION(ripple_animation, zmk_position_state_changed);
```

(which is admittedly rather verbose, but maybe there's a good way to extract some of the boilerplate into some common utility functions for allowing driver instances to dynamically register for events.)